common/reviews/api/node-core-library.api.md

```diff
@@ -26,6 +26,7 @@ export class AlreadyReportedError extends Error {
 // @public
 export class Async {
     static forEachAsync<TEntry>(iterable: Iterable<TEntry> | AsyncIterable<TEntry>, callback: (entry: TEntry, arrayIndex: number) => Promise<void>, options?: IAsyncParallelismOptions | undefined): Promise<void>;
+    static forEachWeightedAsync<TEntry extends IWeightedIterable>(iterable: Iterable<TEntry> | AsyncIterable<TEntry>, callback: (entry: TEntry, arrayIndex: number) => Promise<void>, options?: IAsyncParallelismOptions | undefined): Promise<void>;
```

Comment:
Instead of adding a second function, you add a `weighted?: boolean` option to `IAsyncParallelismOptions`, and if that's set to `true`, take the weight from the array elements.

The function signature would look like this:

```TypeScript
    public static forEachAsync<TEntry>(iterable: Iterable<TEntry> | AsyncIterable<TEntry>, callback: (entry: TEntry, arrayIndex: number) => Promise<void>, options?: IAsyncParallelismOptions): Promise<void>;
    public static forEachAsync<TEntry extends IWeightedIterable>(iterable: Iterable<TEntry> | AsyncIterable<TEntry>, callback: (entry: TEntry, arrayIndex: number) => Promise<void>, options: IAsyncParallelismOptions & { weighted: true }): Promise<void>;
    public static forEachAsync<TEntry>(iterable: Iterable<TEntry> | AsyncIterable<TEntry>, callback: (entry: TEntry, arrayIndex: number) => Promise<void>, options?: IAsyncParallelismOptions): Promise<void> {
      // implementation ...
    }
```

Reply:
This was an interesting challenge, TS overloading is definitely finnicky. I think it's in a good place though now in [06c7679](https://github.com/microsoft/rushstack/pull/4672/commits/06c7679f01b9f9d6267bd9d580f9c401c62ddbb5).